gno.land/cmd/genesis/balances_add.go

```diff
@@ -0,0 +1,346 @@
+
+	// Get the balance sheet from the source
+	switch {
+	case singleEntriesSet:
+		balances, err = getBalancesFromEntries(cfg.singleEntries)
+	case balanceSheetSet:
+		// Open the balance sheet
+		file, loadErr := os.Open(cfg.balanceSheet)
+		if loadErr != nil {
+			return fmt.Errorf("unable to open balance sheet, %w", loadErr)
+		}
+
+		balances, err = getBalancesFromSheet(file)
+	default:
+		// Open the transactions file
+		file, loadErr := os.Open(cfg.parseExport)
+		if loadErr != nil {
+			return fmt.Errorf("unable to open transactions file, %w", loadErr)
+		}
+
+		balances, err = getBalancesFromTransactions(ctx, file)
+	}
+
+	if err != nil {
+		return fmt.Errorf("unable to get balances, %w", err)
+	}
+
+	// Initialize genesis app state if it is not initialized already
+	if genesis.AppState == nil {
+		genesis.AppState = gnoland.GnoGenesisState{}
+	}
+
+	// Construct the initial genesis balance sheet
+	state := genesis.AppState.(gnoland.GnoGenesisState)
+	genesisBalances, err := extractGenesisBalances(state)
+	if err != nil {
+		return err
+	}
+
+	// Merge the two balance sheets, with the input
+	// having precedence over the genesis transactions
+	balances.leftMerge(genesisBalances)
+
+	// Save the balances
+	state.Balances = balances.toList()
+	genesis.AppState = state
+
+	// Save the updated genesis
+	if err := genesis.SaveAs(cfg.rootCfg.genesisPath); err != nil {
+		return fmt.Errorf("unable to save genesis.json, %w", err)
+	}
+
+	io.Printfln(
+		"%d pre-mines saved",
+		len(balances),
+	)
+
+	io.Println()
+
+	for address, balance := range balances {
+		io.Printfln("%s:%dugnot", address.String(), balance)
+	}
+
+	return nil
+}
+
+// validateSetModes validates a good mode was
+// set for the balance addition
+func validateSetModes(modes []bool) error {
+	anySet := false
+
+	for _, mode := range modes {
+		if !mode {
+			continue
+		}
+
+		if anySet {
+			return errMultipleBalanceSources
+		}
+
+		anySet = true
+	}
+
+	if !anySet {
+		return errNoBalanceSource
+	}
+
+	return nil
+}
+
+// getBalancesFromEntries extracts the balance entries
+// from the array of balance
+func getBalancesFromEntries(entries []string) (accountBalances, error) {
+	balances := make(accountBalances)
+
+	for _, entry := range entries {
+		accountBalance, err := getBalanceFromEntry(entry)
+		if err != nil {
+			return nil, fmt.Errorf("unable to extract balance data, %w", err)
+		}
+
+		balances[accountBalance.address] = accountBalance.amount
+	}
+
+	return balances, nil
+}
+
+// getBalancesFromSheet extracts the balance sheet from the passed in
+// balance sheet file, that has the format of <address>=<amount>ugnot
+func getBalancesFromSheet(sheet io.Reader) (accountBalances, error) {
+	// Parse the balances
+	balances := make(accountBalances)
+	scanner := bufio.NewScanner(sheet)
+
+	for scanner.Scan() {
+		entry := scanner.Text()
+
+		// Remove comments
+		entry = strings.TrimSpace(entry)
+		entry = strings.Split(entry, "#")[0]
+		entry = strings.TrimSpace(entry)
```

Comment:
maybe you can trim only once at the end.

What happens if you have several comment lines?

Reply:
> maybe you can trim only once at the end.

Good catch, this code was taken from `gnoland start`. Dropped the initial trim:

8df30bc

> What happens if you have several comment lines?

This format is not standard or anything like that. I just adapted the one we already use (`gno/gno.land/genesis/genesis_balances.txt`). It can only have single-line comments using `#` (or rather, _expects_ to have them in this format). The parsing errors out if it encounters anything funky